### PR TITLE
Tools: Testbench: Add xt-testbench build for mtl

### PR DESCRIPTION
**Pull request by singalsu**
See commits for changes descriptions

### CODE REVIEW

File: scripts/set_xtensa_params.sh [hunk 1]
```diff
@@ -81,4 +81,11 @@ case "$platform" in
 	HOST="xtensa-mt8195-elf"
 	XTENSA_TOOLS_VERSION="RI-2019.1-linux"
 	;;
+    mtl)
+	PLATFORM="mtl"
+	XTENSA_CORE="ace10_LX7HiFi4_2022_10"
+	XTENSA_TOOLS_VERSION="RI-2022.10-linux"
+	COMPILER="xt-clang"
+	LDEXTRAFLAGS=
+	;;
```
Comment (singalsu) [inline scripts/set_xtensa_params.sh:90]: @marc-hb Do you have ideas how to clean-up set_xtensa_params.sh. I guess this was duplicated from zephyr build when it was still a shell script.

File: scripts/rebuild-testbench.sh [hunk 2]
```diff
@@ -56,6 +56,8 @@ setup_xtensa_tools_build()
     BUILD_TYPE=xt
     BUILD_TARGET=
     BUILD_DIR_NAME=build_xt_testbench
+    COMPILER="xt-xcc"
+    LDEXTRAFLAGS="-Wl,-LE"
```
Comment (marc-hb) [inline scripts/rebuild-testbench.sh:60]: This flag never had any effect because there is no `./E/` directory. I'm dropping it in #8384

File: tools/tplg_parser/CMakeLists.txt [hunk 3]
```diff
@@ -64,7 +64,7 @@ endif()
 
 # TODO: add IPC4 option when it's ready.
 target_compile_options(sof_tplg_parser PRIVATE
-	-g -O -Wall -Werror -Wl,-EL -fPIC -DPIC
+	-g -O -Wall -Werror -fPIC -DPIC
```
Comment (marc-hb) [inline tools/tplg_parser/CMakeLists.txt:67]: Do you remember why you specified the endianness in commit 0c35f1efde17bbd27d?

File: scripts/rebuild-testbench.sh (hunk 2 above)
Comment (singalsu) [reply scripts/rebuild-testbench.sh:60]: (facepalm) good that you spotted it, I was yesterday trying to find from manual what it's doing :)

File: scripts/set_xtensa_params.sh (hunk 1 above)
Comment (singalsu) [reply scripts/set_xtensa_params.sh:90]: I think there's a lot of unused environment variables like HOST. Or at least MTL built without setting it, but maybe it's for gcc crosscompiler for other platforms. Currently testbench build is not supporting it.

File: tools/tplg_parser/CMakeLists.txt (hunk 3 above)
Comment (singalsu) [reply tools/tplg_parser/CMakeLists.txt:67]: No, I don't. I have probably copied it from somewhere else from that time SOF, and it looks totally unnecessary. Thanks for checking this draft!

File: src/include/sof/trace/preproc.h [hunk 4]
```diff
@@ -158,8 +158,10 @@
 		return_t META_CONCAT(prefix, postfix) (args)
 
 /* counteract compiler warning about unused variables */
+#ifndef UNUSED
 #define UNUSED(arg1, ...) do { META_RECURSE( \
 	META_MAP_AGGREGATE(1, _META_VOID2, _META_VOID(arg1), __VA_ARGS__)); \
 	} while (0)
+#endif
```
Comment (singalsu) [inline src/include/sof/trace/preproc.h:165]: I feel this change is not good. Would it be better to rename UNUSED() to SOF_UNUSED() or SOF_TRACE_UNUSED()?

**Review by lgirdwood — APPROVED** [review]
I guess you can deal with teh UNUSED macro name in a subsequent PR ?

File: src/include/sof/trace/preproc.h (hunk 4 above)
Comment (marc-hb) [reply src/include/sof/trace/preproc.h:165]: Yes please rename this to SOF_TRACE_UNUSED() in a separate commit (or even separate PR).

> I Think the pre-processing code is now only used for logging, could be deprecated...

... for Intel.

File: tools/testbench/testbench.c [hunk 5]
```diff
@@ -353,13 +353,14 @@ static int parse_input_args(int argc, char **argv, struct testbench_prm *tp)
...
-			ret = -EINVAL;
-			/* fallthrough */
 		case 'h':
 			print_usage(argv[0]);
 			exit(EXIT_SUCCESS);
+
+		default:
+			fprintf(stderr, "unknown option %c\n", option);
+			print_usage(argv[0]);
+			ret = -EINVAL;
```
Comment (marc-hb) [inline tools/testbench/testbench.c:363]: You're getting away with this problem with a refactoring this time around but it's "selfish"! How about other statements elsewhere in the code?

Could you try to find something portable across both gcc and clang?
https://stackoverflow.com/questions/45129741/gcc-7-wimplicit-fallthrough-warnings-and-portable-way-to-clear-them

**Review by marc-hb — DISMISSED** [review]
SOF_TRACE_UNUSED

File: tools/testbench/testbench.c (hunk 5 above)
Comment (singalsu) [reply tools/testbench/testbench.c:363]: I'll try if there's something that works for all compilers with help of that link (yesterday I didn't find).

File: src/include/sof/trace/preproc.h [hunk 6]
```diff
@@ -158,7 +158,7 @@
 		return_t META_CONCAT(prefix, postfix) (args)
 
 /* counteract compiler warning about unused variables */
-#define UNUSED(arg1, ...) do { META_RECURSE( \
```
Comment (singalsu) [inline src/include/sof/trace/preproc.h:161]: @marc-hb  I added this separately as #8395 so I will remove it this commit.

File: tools/testbench/testbench.c (hunk 5 above)
Comment (marc-hb) [reply tools/testbench/testbench.c:363]: Can you please try something like this? It supposedly works with both gcc and clang.

```C
#if defined __has_cpp_attribute && __has_cpp_attribute(fallthrough)
#  define SOF_FALLTHRU __attribute__((__fallthrough__)
#elif <FUTURE_UNUSUAL_COMPILER>
#  define SOF_FALLTHRU <something unusual>
#else
#  warning unknown compiler, undefined SOF_FALLTHRU
#endif

```

Comment (marc-hb) [reply tools/testbench/testbench.c:363]: https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/toolchain uses `__attribute__((__fallthrough__)` for both gcc and clang.

File: src/include/sof/trace/trace.h [hunk 7]
```diff
@@ -93,7 +93,7 @@ static inline struct trace *trace_get(void)
 
 /* Silences compiler warnings about unused variables */
 #define trace_unused(class, ctx, id_1, id_2, format, ...) \
-	UNUSED(ctx, id_1, id_2, ##__VA_ARGS__)
+	SOF_TRACE_UNUSED(ctx, id_1, id_2, ##__VA_ARGS__)
```
Comment (btian1) [inline src/include/sof/trace/trace.h:96]: please remove this patch, since you submitted with a separate PR.

**Comment by marc-hb** [comment]
`-Wl,-EL` déjà vu:
https://github.com/thesofproject/rimage/commit/bcbcec79e582

File: tools/testbench/testbench.c (hunk 5 above)
Comment (singalsu) [reply tools/testbench/testbench.c:363]: It's used with #define like this:

#ifndef __fallthrough
#if __GNUC__ >= 7
#define __fallthrough        __attribute__((fallthrough))
#else
#define __fallthrough
#endif	/* __GNUC__ >= 7 */
#endif

Comment (singalsu) [reply tools/testbench/testbench.c:363]: I still feel my code re-order is simplest way to address this. The fall-through is unnecessary trickery for a very simple thing.

Comment (singalsu) [reply tools/testbench/testbench.c:363]: Also if you look the original code, there's a bug that I'm fixing the same time:

```
		default:
			fprintf(stderr, "unknown option %c\n", option);
			ret = -EINVAL;
			/* fallthrough */
		case 'h':
			print_usage(argv[0]);
			exit(EXIT_SUCCESS);
		}

		if (ret < 0)
			return ret;

```

The `-EINVAL `is not returned because of `exit(EXIT_SUCCESS)`. Fall-through is not correct here.

File: src/include/sof/trace/trace.h (hunk 7 above)
Comment (singalsu) [reply src/include/sof/trace/trace.h]: Yep, rebase dropped it.

File: tools/testbench/testbench.c (hunk 5 above)
Comment (marc-hb) [reply tools/testbench/testbench.c:363]: > unnecessary trickery for a very simple thing

I never said your change was wrong, I said it was "selfish" and I stand by that. This type of issue is painful to test because you need multiple toolchains and many people use only one. But you already have the environment, skills and a good test case. 

That "trickery" would really not be that complicated (try `git grep META` for some _real_ trickery!) and it would be only in one or two .h files; just one very simple SOF_FALLTHRU keyword everywhere else.


> `#if __GNUC__ >= 7` 

We don't need stuff like that for `testbench.c`, so it could be very simple for now.

Comment (marc-hb) [reply tools/testbench/testbench.c:363]: > Also if you look the original code, there's a bug that I'm fixing the same time: [...]
> Fall-through is not correct here.

The current commit message still refers to fall through only.

Comment (singalsu) [reply tools/testbench/testbench.c:363]: @marc-hb I just updated the commit text. Is it OK now?

Comment (marc-hb) [reply tools/testbench/testbench.c:363]: Sure. Just for the record I never formally "requested" fallthrough changes :-)

**Review by marc-hb — APPROVED** [review]
